File: apps/backend/src/db/migrate.ts
import 'dotenv/config';
import { readdir, readFile } from 'node:fs/promises';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { pool } from './client.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const MIGRATIONS_DIR = join(__dirname, 'migrations');

async function run() {
  const client = await pool.connect();
  try {
    await client.query(`
      CREATE TABLE IF NOT EXISTS migrations (
        name   TEXT        PRIMARY KEY,
        run_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
      )
    `);

    const { rows: applied } = await client.query<{ name: string }>(
      'SELECT name FROM migrations ORDER BY name'
    );
    const appliedSet = new Set(applied.map((r) => r.name));

    const files = (await readdir(MIGRATIONS_DIR))
      .filter((f) => f.endsWith('.sql'))
      .sort();

    let ran = 0;
    for (const file of files) {
      if (appliedSet.has(file)) {
        console.log(`[migrate] skip  ${file}`);
        continue;
      }

      const sql = await readFile(join(MIGRATIONS_DIR, file), 'utf8');

      await client.query('BEGIN');
      try {
        await client.query(sql);
        await client.query('INSERT INTO migrations (name) VALUES ($1)', [file]);
        await client.query('COMMIT');
        console.log(`[migrate] ran   ${file}`);
        ran++;
      } catch (err) {
        await client.query('ROLLBACK');
        throw new Error(
          `Migration ${file} failed: ${(err as Error).message}`
        );
      }
    }

    console.log(`[migrate] done — ${ran} migration(s) applied.`);
  } finally {
    client.release();
    await pool.end();
  }
}

run().catch((err) => {
  console.error('[migrate] Fatal:', err.message);
  process.exit(1);
});

File: backend/app/db/migrate.py
"""Migration runner — port of `src/db/migrate.ts`.

Run as a module:
    python -m app.db.migrate

Reads SQL files from app/db/migrations/, applies them in lexicographic order,
each inside its own transaction, tracking applied files in a `migrations`
table. Idempotent: re-running is safe.
"""

from __future__ import annotations

import asyncio
import sys
from pathlib import Path

import asyncpg

from ..config import settings

MIGRATIONS_DIR = Path(__file__).resolve().parent / "migrations"


async def run() -> int:
    conn = await asyncpg.connect(dsn=settings.database_url)
    try:
        await conn.execute("SELECT pg_advisory_lock(80671001)")
        await conn.execute(
            """
            CREATE EXTENSION IF NOT EXISTS pgcrypto;

            CREATE TABLE IF NOT EXISTS migrations (
              name   TEXT        PRIMARY KEY,
              run_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
            )
            """
        )

        applied_rows = await conn.fetch("SELECT name FROM migrations ORDER BY name")
        applied = {r["name"] for r in applied_rows}

        files = sorted(p.name for p in MIGRATIONS_DIR.glob("*.sql"))

        ran = 0
        for file in files:
            if file in applied:
                print(f"[migrate] skip  {file}")
                continue

            sql = (MIGRATIONS_DIR / file).read_text(encoding="utf-8")

            async with conn.transaction():
                await conn.execute(sql)
                await conn.execute("INSERT INTO migrations (name) VALUES ($1)", file)
            print(f"[migrate] ran   {file}")
            ran += 1

        print(f"[migrate] done — {ran} migration(s) applied.")
        return 0
    finally:
        try:
            await conn.execute("SELECT pg_advisory_unlock(80671001)")
        except Exception:
            pass
        await conn.close()


def main() -> None:
    try:
        sys.exit(asyncio.run(run()))
    except Exception as e:
        print(f"[migrate] Fatal: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
